sql/binlog_ostream.cc

```diff
@@ -105,12 +105,21 @@ bool IO_CACHE_binlog_cache_storage::begin(unsigned char **buffer,
     my_error(ER_ERROR_ON_WRITE, MYF(MY_WME), tmp_file_name(), errno, errbuf);
     return true;
   }
+
+  if (my_b_bytes_in_cache(&m_io_cache) > 0) {
+    *buffer = m_io_cache.read_pos;
+    *length = my_b_bytes_in_cache(&m_io_cache);
+
+    m_io_cache.read_pos = m_io_cache.read_end;
+    return false;
+  }
+
   return next(buffer, length);
 }
 
 bool IO_CACHE_binlog_cache_storage::next(unsigned char **buffer,
                                          my_off_t *length) {
-  my_b_fill(&m_io_cache);
+  if (my_b_bytes_in_cache(&m_io_cache) == 0) my_b_fill(&m_io_cache);
```

Comment:
In the description of the commit b2a8ee5
I found the following comment
```
'Transaction_Message::append_cache()' method could call 'my_b_fill()' in case
when 'info->read_pos != info->read_end' which is not supported in the
encrypted mode.
```
So, I believe `my_b_bytes_in_cache(&m_io_cache) == 0`  (which is `*m_io_cache->current_end == *m_io_cache->current_pos`) should work the same way as `m_io_cache->read_pos == m_io_cache->read_end` here.

My only concern is why do we need modifications in the `IO_CACHE_binlog_cache_storage::tmp_file_name()`. To me, just calling the modified `next()` from there should be enough.